rust-symcrypt/src/rsa/mod.rs

```diff
@@ -0,0 +1,716 @@
+/// This is to maintain interop with legacy Windows code and does not enforce any checks under the covers. The caller is responsible for ensuring the key is used correctly.
+pub enum RsaKeyUsage {
+    // When using [`RsaKeyUsage::Sign`], the intended usage for the key will be only Signing.
+    Sign,
+
+    /// When using [`RsaKeyUsage::Encrypt`], the intended usage for the key will be only Encryption.
+    Encrypt,
+
+    /// When using [RsaKeyUsage::SignAndEncrypt`], the intended usage for the key can be either Signing or Encryption.
+    SignAndEncrypt,
+}
+
+// to_symcrypt_flag converts the RsaKeyUsage to the corresponding SymCrypt flag and only needed internally.
+impl RsaKeyUsage {
+    pub(crate) fn to_symcrypt_flag(&self) -> symcrypt_sys::UINT32 {
+        match self {
+            RsaKeyUsage::Sign => symcrypt_sys::SYMCRYPT_FLAG_RSAKEY_SIGN,
+            RsaKeyUsage::Encrypt => symcrypt_sys::SYMCRYPT_FLAG_RSAKEY_ENCRYPT,
+            RsaKeyUsage::SignAndEncrypt => {
+                symcrypt_sys::SYMCRYPT_FLAG_RSAKEY_ENCRYPT | symcrypt_sys::SYMCRYPT_FLAG_RSAKEY_SIGN
+            }
+        }
+    }
+}
+
+#[derive(Debug)]
+#[allow(dead_code)]
+/// [`RsaKeyPairExportBlob`] holds the values of the `[RsaKeyPair]` when the key pair is exported.  
+/// NOTE: SymCrypt does not pre-append leading 0's if the MSB is set.
+pub struct RsaKeyPairExportBlob {
+    pub modulus: Vec<u8>,
+    pub pub_exp: Vec<u8>,
+    pub p: Vec<u8>,
+    pub q: Vec<u8>,
+    pub d_p: Vec<u8>,
+    pub d_q: Vec<u8>,
+    pub crt_coefficient: Vec<u8>,
+    pub private_exp: Vec<u8>,
+}
+
+#[derive(Debug)]
+#[allow(dead_code)]
+/// [`RsaPublicKeyExportBlob`] holds the values of the `[RsaPublicKey]` when the key is exported.
+/// NOTE: SymCrypt does not pre-append leading 0's if the MSB is set.
+pub struct RsaPublicKeyExportBlob {
+    pub modulus: Vec<u8>,
+    pub pub_exp: Vec<u8>,
+}
+
+/// Impl for RsaKeyPair struct.
+impl RsaKeyPair {
+    /// `generate_new()` generates a random Rsa key based on the provided parameters.
+    ///
+    /// `n_bits_mod` represents a `u32` that is the desired bit length of the Rsa key, `n_bits_mod` must be at least 1024 bits.
+    ///
+    ///
+    /// `pub_exp` takes in an `Option<u64>` that is the public exponent. If `None` is provided, the default `2^16 +1` will be used.
+    ///  
+    /// `key_usage` takes in a [`RsaKeyUsage`] and will indicate if this key will be used for [`RsaKeyUsage::Sign`], or [`RsaKeyUsage::Encrypt`], or [`RsaKeyUsage::SignAndEncrypt`]
+    pub fn generate_new(
+        n_bits_mod: u32,
+        pub_exp: Option<&[u8]>,
+        key_usage: RsaKeyUsage,
+    ) -> Result<Self, SymCryptError> {
+        let (pub_exp_ptr, pub_exp_count) = match pub_exp {
+            Some(exp) => {
+                let u64_pub_exp = load_msb_first_u64(exp)?;
```

Comment:
If you require the caller to pass in a pub_exp that is u64 value, you will need to provide a public function for this and clearly document and provide example code for.